tests/testthat/test-ggplot-polygons.R

```diff
@@ -1,115 +1,141 @@
 context("polygon")
 
-test_that("filled polygons become several traces", {
-  poly.df <- data.frame(x=c(0, 1, 1, 0, 2, 3, 3, 2)+10,
-                        y=c(0, 0, 1, 1, 0, 0, 1, 1)+10,
-                        g=c(1, 1, 1, 1, 2, 2, 2, 2))
-  poly.df$lab <- paste0("name", poly.df$g)
+expect_traces <- function(gg, n.traces, name){
+  stopifnot(is.ggplot(gg))
+  stopifnot(is.numeric(n.traces))
+  save_outputs(gg, paste0("polygon-", name))
+  L <- gg2list(gg)
+  is.trace <- names(L) == ""
+  all.traces <- L[is.trace]
+  no.data <- sapply(all.traces, function(tr) {
+    is.null(tr[["x"]]) && is.null(tr[["y"]])
+  })
+  has.data <- all.traces[!no.data]
+  expect_equal(length(has.data), n.traces)
+  list(traces=has.data, kwargs=L$kwargs)
+}
+
+poly.df <- data.frame(x=c(0, 1, 1, 0, 2, 3, 3, 2)+10,
+                      y=c(0, 0, 1, 1, 0, 0, 1, 1),
+                      g=c(1, 1, 1, 1, 2, 2, 2, 2),
+                      lab=rep(c("left", "right"), each=4))
+
+test_that("polygons filled with the same color become one trace", {
   gg <- ggplot(poly.df)+
     geom_polygon(aes(x, y, group=g))
-  info <- gg2list(gg)
-  expect_equal(length(info), 3)
-  expect_equal(info[[1]]$x, c(10, 11, 11, 10, 10))
-  expect_equal(info[[1]]$y, c(10, 10, 11, 11, 10))
-  expect_equal(info[[2]]$x, c(12, 13, 13, 12, 12))
-  expect_equal(info[[2]]$y, c(10, 10, 11, 11, 10))
-  expect_identical(info[[1]]$line$color, "transparent")
-  expect_identical(info[[2]]$line$color, "transparent")
-  
-  expect_identical(info[[1]]$showlegend, FALSE)
-  expect_identical(info[[2]]$showlegend, FALSE)
-
-  save_outputs(gg, "polygons-filled-polygons")
-
-  first.color <- rgb(0.23, 0.45, 0.67)
-  gg <- ggplot(poly.df)+
-    geom_polygon(aes(x, y, color=lab), fill="grey")+
-    scale_color_manual(values=c(name1=first.color, name2="springgreen3"))
-  info <- gg2list(gg)
-  expect_equal(length(info), 3)
-  expect_equal(info[[1]]$x, c(10, 11, 11, 10, 10))
-  expect_equal(info[[1]]$y, c(10, 10, 11, 11, 10))
-  expect_equal(info[[1]]$fillcolor, toRGB("grey"))
-  expect_equal(info[[1]]$line$color, toRGB(first.color))
-  expect_equal(info[[1]]$name, "name1")
-  expect_equal(info[[2]]$x, c(12, 13, 13, 12, 12))
-  expect_equal(info[[2]]$y, c(10, 10, 11, 11, 10))
-  expect_equal(info[[2]]$fillcolor, toRGB("grey"))
-  expect_equal(info[[2]]$line$color, toRGB("springgreen3"))
-  expect_equal(info[[2]]$name, "name2")
-
-  expect_identical(info[[1]]$showlegend, TRUE)
-  expect_identical(info[[2]]$showlegend, TRUE)
+  info <- expect_traces(gg, 1, "black")
```

Comment:
I updated the tests to check for the NA values and retracing the first points of each group.